file.go

```diff
@@ -0,0 +1,149 @@
+package gluster
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/gluster/gogfapi/gfapi"
+	billy "gopkg.in/src-d/go-billy.v4"
+)
+
+var (
+	_ billy.File = new(File)
+)
+
+type mode int
+
+const (
+	// ErrReadOnly returned when file is read only.
+	ErrReadOnly = "cannot write in %s, the file is read only"
+	// ErrWriteOnly returned when file is write only.
+	ErrWriteOnly = "cannot read from %s, the file is write only"
+
+	read  mode = 0
+	write mode = 1
+)
+
+// File holds a gluster file descritor.
+type File struct {
+	path  string
+	g     *gfapi.File
+	flags int
+}
+
+// NewFile creates a new wrapped gluster file descriptor.
+func NewFile(name string, g *gfapi.File, flags int) *File {
+	return &File{path: name, g: g, flags: flags}
+}
+
+// Name implements billy.File interface.
+func (f *File) Name() string {
+	return f.path
+}
+
+// Write implements billy.File interface.
+func (f *File) Write(p []byte) (int, error) {
+	err := f.checkFlags(write)
+	if err != nil {
+		return 0, err
+	}
+
+	return f.g.Write(p)
+}
+
+// Read implements billy.File interface.
+func (f *File) Read(p []byte) (int, error) {
+	err := f.checkFlags(read)
+	if err != nil {
+		return 0, err
+	}
+
+	n, err := f.g.Read(p)
+	// on error n is negative
```

Comment:
it doesn't set any `errno` which we can read and convert to error?

Reply:
It is done already by cgo when calling the C function. The problem is that `n` is not cleared by the library and some functions complain when the number of bytes read are negative, even when there is an error.

https://github.com/gluster/gogfapi/blob/master/gfapi/fd.go#L93-L100